### PR TITLE
feat: add PoolTogether plugin

### LINKED ISSUE #3343: [Bounty] Integrate Pooltogether, a savings game, in Mask Plugin
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks
- Bounty Size
  - [x] 400 $MASK or $2000

## Part 1: Problem Description

**Background**
Pooltogether is a defi savings game where part of the interest is put in a lottery pot and all depositers have a chance (propotional to the savings amount) to win the lotttery. We think it has really good synergy with traditional social networks and thus worth the integration. 

Previously defi plugins on Mask, like the trading widgets powered by 0x and Uniswap, has been plagued by high ETH gas prices.  However with the launch of polygon and BSC integration, the problem should have been largely resolved. Savings app which was hindered by monstrous ETH gas price is now very much affordable. 

Pooltogether is on both eth mainnet and polygon with identical UI.

**Solution**
Create a plugin that integrate pooltogether.  The UI could resemble Dhedge in a card style with different tabs serving different purposes. 

image
_Dhedge Plugin UI_

image
_Pooltogether pool list_

image
_Pooltogether deposit page_

Default page could be the pool list with return, time remaining and other basic info like shown in the official website (IMO it's clean enough for porting directly) . Once click deposit, users will be taken to deposit page where odds are displayed real time

A second tab would be my page where all my lottery bets are displayed

Anylink from pooltogether on Twitter could trigger the plugin ( could have some slightly different treatment). The tab order can be changed, corresponding landing and ending page needs to be added. Share button for such a social game is a must. 

Language like 
**I just deposit xxx dollar into xxx pool, can I win the lottery this week?** 
should be added to help spread the word.

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- https://pooltogether.com/
- https://docs.pooltogether.com/

## Part 3: Contact and Support
@yisiliu for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q
Pooltogether discord server: https://discord.com/invite/hxPhPDW

**Submission Requirements**
Please include demo to show the integration is working.

**Winner Announcement Date**
Winners will be announced at the conclusion of Gitcoin 10 Hackathon

**Judging Criteria**
Have a working demo.
Clean code & tech design
Clear UI, easy to use and generally good user experience

### PR DESCRIPTION
**Pull request by iRhonin**
closes #3343

https://user-images.githubusercontent.com/13151232/124670112-bf586980-dec8-11eb-9784-bc618fa7ac09.mp4

### CODE REVIEW

File: packages/maskbook/src/plugins/PluginSerivce.ts [hunk 1]
```diff
@@ -9,3 +9,4 @@
 
 // Please make sure you have registered your plugin UI at ./PluginUI
 import './dHEDGE/messages'
+import './PoolTogether/messages'
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PluginSerivce.ts:12]: @Jack-Works Do we have any document for the new plugin infra?

File: packages/maskbook/src/plugins/PoolTogether/UI/Account.tsx [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            color: '#ffffff',
+        },
+    },
+    pools: {
+        margin: theme.spacing(1, 0),
+    },
+}))
+
+interface AccountProps {
+    pools: [Pool]
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/UI/Account.tsx:56]: Do you mean `Pool[]`?

File: packages/maskbook/src/plugins/PoolTogether/UI/Account.tsx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+export function Account(props: AccountProps) {
+    const { pools } = props
+
+    const { t } = useI18N()
+    const classes = useStyles()
+
+    const { value: balances = [], loading, retry, error } = useAccountBalance(pools)
+    console.log(balances)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/UI/Account.tsx:66]: Remove it.

File: packages/maskbook/src/plugins/PoolTogether/UI/Account.tsx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+    if (error) {
+        return <RefreshIcon className={classes.refresh} color="primary" onClick={retry} />
+    }
+
+    const noZeroBalances = balances.filter((balance) => parseInt(balance.account.ticketBalance) !== 0)
+    const totalUsdBalance = noZeroBalances
+        .map((balance) => {
+            const ticketBalance = parseFloat(
+                formatBalance(balance.account.ticketBalance, parseInt(balance.pool.tokens.ticket.decimals)),
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/UI/Account.tsx:80]:
```suggestion
                formatBalance(balance.account.ticketBalance, Number.parseInt(balance.pool.tokens.ticket.decimals)),
```

File: packages/maskbook/src/plugins/PoolTogether/UI/AccountPool.tsx [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    const { accountPool } = props
+    const token = accountPool.pool.tokens.underlyingToken
+
+    const { t } = useI18N()
+    const classes = useStyles()
+
+    const poolURL = useManagePoolURL(accountPool.pool)
+    const chainId = useChainId()
+
+    const formatedBalance = parseFloat(
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/UI/AccountPool.tsx:117]:
```suggestion
    const formatedBalance = Number.parseFloat(
```

File: packages/maskbook/src/plugins/PoolTogether/UI/AccountPool.tsx [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        formatBalance(accountPool.account.ticketBalance, parseInt(accountPool.pool.tokens.ticket.decimals)),
+    ).toLocaleString(undefined, {
+        maximumFractionDigits: 6,
+    })
+
+    const odds = calculateOdds(
+        parseFloat(formatedBalance),
+        parseFloat(accountPool.pool.tokens.ticket.totalSupply),
+        parseInt(accountPool.pool.config.numberOfWinners),
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/UI/AccountPool.tsx:126]:
```suggestion
        Number.parseFloat(formatedBalance),
        Number.parseFloat(accountPool.pool.tokens.ticket.totalSupply),
        Number.parseInt(accountPool.pool.config.numberOfWinners),
```

File: packages/maskbook/src/plugins/PoolTogether/UI/AccountPool.tsx [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    const token = accountPool.pool.tokens.underlyingToken
+
+    const { t } = useI18N()
+    const classes = useStyles()
+
+    const poolURL = useManagePoolURL(accountPool.pool)
+    const chainId = useChainId()
+
+    const formatedBalance = parseFloat(
+        formatBalance(accountPool.account.ticketBalance, parseInt(accountPool.pool.tokens.ticket.decimals)),
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/UI/AccountPool.tsx:118]:
```suggestion
        formatBalance(accountPool.account.ticketBalance, Number.parseInt(accountPool.pool.tokens.ticket.decimals)),
```

File: packages/maskbook/src/plugins/PoolTogether/UI/CountdownView.tsx [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+export const CountdownView = (props: CountdownProps) => {
+    const classes = useStylesExtends(useStyles(), props)
+
+    const { secondsRemaining, msgOnEnd } = props
+    const [secs, setSecs] = useState(secondsRemaining)
+
+    // useInterval(() => {
+    //     setSecs(secs - 1)
+    // }, ONE_SECOND)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/UI/CountdownView.tsx:55]:
```suggestion
```

File: packages/maskbook/src/plugins/PoolTogether/UI/DepositDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+    } = useTokenBalance(token?.type ?? EthereumTokenType.Native, token?.address ?? '')
+    //#endregion
+
+    useEffect(() => {
+        setOdds(
+            pool
+                ? calculateOdds(
+                      parseFloat(rawAmount),
+                      parseFloat(pool.tokens.ticket.totalSupply),
+                      parseInt(pool.config.numberOfWinners),
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/UI/DepositDialog.tsx:148]:
```suggestion
                      Number.parseFloat(rawAmount),
                      Number.parseFloat(pool.tokens.ticket.totalSupply),
                      Number.parseInt(pool.config.numberOfWinners),
```

File: packages/maskbook/src/plugins/PoolTogether/types.tsx [hunk 10]
```diff
@@ -0,0 +1,336 @@
+export interface Ticket {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/types.tsx:1]: Add a type for 

```typescript
{
     hex: string
     type: string
}
```

File: packages/maskbook/src/plugins/PluginSerivce.ts (hunk 1 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/PluginSerivce.ts:12]: All APIs are well documented in `packages\plugin-infra\src\types.ts` and the plugin structure can be seen in `packages\plugins\example`.

File: packages/maskbook/src/plugins/PoolTogether/README.md [hunk 11]
```diff
@@ -0,0 +1,22 @@
+# Plugin: PoolTogether
+
+This plugin fetch and show [PoolTogether](https://pooltogether.com/) pools data and allow user to invest on that poll and see her/his account data.
+
+## Feature Set
+
+- [x] Detect PoolTogether address
+- [x] Show pools list
+- [x] Deposit
+- [x]] My account tab
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/README.md:10]:
```suggestion
- [x] My account tab
```

File: packages/maskbook/src/plugins/PoolTogether/README.md [hunk 12]
```diff
@@ -0,0 +1,22 @@
+# Plugin: PoolTogether
+
+This plugin fetch and show [PoolTogether](https://pooltogether.com/) pools data and allow user to invest on that poll and see her/his account data.
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/README.md:3]:
```suggestion
This plugin fetch and show [PoolTogether](https://pooltogether.com) pools data and allow user to invest on that poll and see her/his account data.
```

File: packages/maskbook/src/plugins/PoolTogether/README.md [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+## Feature Set
+
+- [x] Detect PoolTogether address
+- [x] Show pools list
+- [x] Deposit
+- [x]] My account tab
+
+## Related projects
+
+- <https://pooltogether.com/>
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/README.md:14]:
```suggestion
- <https://pooltogether.com>
```

**Comment by iRhonin** [comment]
Hi @guanbinrui and @septs, 

Thank you for the detailed review,

All of them are resolved now.

File: packages/maskbook/src/plugins/PoolTogether/UI/AccountPool.tsx [hunk 14]
```diff
@@ -114,16 +114,16 @@ export function AccountPool(props: AccountPoolProps) {
     const poolURL = useManagePoolURL(accountPool.pool)
     const chainId = useChainId()
 
-    const formatedBalance = parseFloat(
-        formatBalance(accountPool.account.ticketBalance, parseInt(accountPool.pool.tokens.ticket.decimals)),
+    const formatedBalance = Number.parseFloat(
+        formatBalance(accountPool.account.ticketBalance, Number.parseInt(accountPool.pool.tokens.ticket.decimals)),
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/UI/AccountPool.tsx:118]:
```suggestion
        formatBalance(accountPool.account.ticketBalance, Number.parseInt(accountPool.pool.tokens.ticket.decimals, 10)),
```

File: packages/maskbook/src/plugins/PoolTogether/UI/PoolView.tsx [hunk 15]
```diff
@@ -177,7 +177,7 @@ export function PoolView(props: PoolProps) {
     //#endregion
 
     //#region process data
-    const prizePeriodSeconds = parseInt(pool.config.prizePeriodSeconds)
+    const prizePeriodSeconds = Number.parseInt(pool.config.prizePeriodSeconds)
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/UI/PoolView.tsx:180]:
```suggestion
    const prizePeriodSeconds = Number.parseInt(pool.config.prizePeriodSeconds, 10)
```

**Review by septs — COMMENTED** [review]
plz resolve warning

**Comment by iRhonin** [comment]
> plz resolve warning

image
this one?

File: packages/maskbook/src/plugins/PluginSerivce.ts (hunk 1 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/PluginSerivce.ts:12]: Please remove this file.

File: packages/maskbook/src/plugins/PoolTogether/messages.ts [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    | {
+          open: false
+      }
+
+interface PoolTogetherMessages {
+    DepositDialogUpdated: DepositDialogUpdated
+
+    rpc: unknown
+}
+
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/PoolTogether/messages.ts:23]:
```suggestion
import.meta.webpackHot?.accept()
```

**Comment by iRhonin** [comment]
The branch is now up-to-date and Mask Pool is implemented. Ready for review and merge.

File: packages/maskbook/src/plugins/PluginSerivce.ts (hunk 1 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/PluginSerivce.ts:12]: @iRhonin hi please remove this file.

Comment (iRhonin) [reply packages/maskbook/src/plugins/PluginSerivce.ts]: Thanks @Jack-Works, done.

File: packages/maskbook/src/plugins/PoolTogether/UI/NetworkVIew.tsx [hunk 17]
```diff
@@ -0,0 +1,38 @@
+import { ChainId, getChainDetailed, getNetworkTypeFromChainId } from '@masknet/web3-shared'
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/UI/NetworkVIew.tsx:1]: plz rename this file, `NetworkVIew.tsx` to `NetworkView.tsx`

File: packages/maskbook/src/plugins/PoolTogether/constants.ts [hunk 18]
```diff
@@ -0,0 +1,18 @@
+import { escapeRegExp } from 'lodash-es'
+
+export const POOLTOGETHER_PLUGIN_ID = 'co.pooltogether'
+
+export const BASE_URL = 'pooltogether.com'
+export const APP_URL = 'https://app.pooltogether.com'
+export const API_URL = 'https://pooltogether-api.com/'
+export const COMMINUTY_URL = 'https://community.pooltogether.com/'
+
+export const URL_PATTERN = new RegExp(`(http|https)\:\/\/.*\.?${escapeRegExp(BASE_URL)}`)
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/constants.ts:10]:
```suggestion
export const URL_PATTERN = new RegExp(`https?\:\/\/.*\.?${escapeRegExp(BASE_URL)}`)
```

File: packages/maskbook/src/plugins/PoolTogether/utils.ts [hunk 19]
```diff
@@ -0,0 +1,94 @@
+import { ChainId } from '@masknet/web3-shared'
+import { ONE_SECOND, ONE_WEEK_SECONDS } from './constants'
+import type { Pool } from './types'
+
+export const subtractDates = (dateA: Date, dateB: Date) => {
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/utils.ts:5]: plz use https://date-fns.org/v2.22.1/docs/sub replace the function.

File: packages/maskbook/src/plugins/PoolTogether/utils.ts [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+    return {
+        days,
+        hours,
+        minutes,
+        seconds,
+    }
+}
+
+export const calculateOdds = (usersTicketBalance: number, totalSupply: number, numberOfWinners: number) => {
+    if (usersTicketBalance === 0 || isNaN(usersTicketBalance)) {
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/utils.ts:45]:
```suggestion
    if (usersTicketBalance === 0 || Number.isNaN(usersTicketBalance)) {
```

File: packages/maskbook/src/plugins/PoolTogether/utils.ts [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+    }
+    return formatedPrize
+}
+
+export const calculateSecondsRemaining = (pool: Pool) => {
+    return (
+        Number.parseInt(pool.prize.prizePeriodStartedAt.hex, 16) +
+        Number.parseInt(pool.prize.prizePeriodSeconds.hex, 16) -
+        new Date().getTime() / ONE_SECOND
+    )
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/utils.ts:81]: why not the parsed value as variable?

```suggestion
    return started_at + seconds - (Date.now() / ONE_SECOND)
```

see https://mdn.io/Date.now

File: packages/maskbook/src/plugins/PoolTogether/utils.ts [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+export const getNetworkColor = (chainId: ChainId) => {
+    switch (chainId) {
+        case ChainId.Mainnet:
+            return '#617fea'
+        case ChainId.Matic:
+            return '#7b41da'
+        // add more if needed
+        default:
+            return '#f1f1f1'
+    }
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/utils.ts:93]: plz early returns.

```suggestion
    }
    return '#f1f1f1'
```

File: packages/web3-shared/src/constants/index.ts [hunk 23]
```diff
@@ -9,6 +9,7 @@ import Trader from '@masknet/web3-constants/data/trader.json'
 import Trending from '@masknet/web3-constants/data/trending.json'
 import URL from '@masknet/web3-constants/data/url.json'
 import RPC from '@masknet/web3-constants/data/rpc.json'
+import Pooltogether from '@masknet/web3-constants/data/pooltogether.json'
```
Comment (septs) [inline packages/web3-shared/src/constants/index.ts:12]:
```suggestion
import PoolTogether from '@masknet/web3-constants/data/pooltogether.json'
```

File: packages/web3-shared/src/constants/index.ts [hunk 24]
```diff
@@ -48,5 +49,8 @@ export const useRPCConstants = hookTransform(getRPCConstants)
 export const getTokenAssetBaseURLConstants = transform(TokenAssetBaseURL)
 export const useTokenAssetBaseURLConstants = hookTransform(getTokenAssetBaseURLConstants)
 
+export const getPooltogetherConstants = transform(Pooltogether)
+export const usePooltogetherConstants = hookTransform(getPooltogetherConstants)
```
Comment (septs) [inline packages/web3-shared/src/constants/index.ts:53]:
```suggestion
export const getPoolTogetherConstants = transform(Pooltogether)
export const usePoolTogetherConstants = hookTransform(getPooltogetherConstants)
```

**Comment by septs** [comment]
@guanbinrui start the pr test? @AinuX

File: packages/maskbook/src/plugins/PoolTogether/UI/PoolView.tsx [hunk 25]
```diff
@@ -0,0 +1,299 @@
...
+}
+
+export function PoolView(props: PoolProps) {
+    const { pool } = props
+    const classes = useStyles()
+    const { t } = useI18N()
+
+    const poolURL = usePoolURL(pool)
+    const chainId = useChainId()
+    const [prize, setPrize] = useState('TBD')
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/PoolTogether/UI/PoolView.tsx:169]: 👀

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/plugins/PoolTogether/constants.ts (hunk 18 above)
Comment (iRhonin) [reply packages/maskbook/src/plugins/PoolTogether/constants.ts:10]: @Jack-Works can you please explain how? 
URL should match on any subdomian.

File: packages/maskbook/src/plugins/PoolTogether/UI/PoolView.tsx (hunk 25 above)
Comment (iRhonin) [reply packages/maskbook/src/plugins/PoolTogether/UI/PoolView.tsx:169]: I did this to set a default value for the prize, is this not a good way?

File: packages/maskbook/src/plugins/PoolTogether/utils.ts (hunk 19 above)
Comment (iRhonin) [reply packages/maskbook/src/plugins/PoolTogether/utils.ts:5]: I rewrote this function, now it takes `seconds` and returns `{days, hours, minutes, seconds}`.

File: packages/maskbook/src/plugins/PoolTogether/constants.ts (hunk 18 above)
Comment (iRhonin) [reply packages/maskbook/src/plugins/PoolTogether/constants.ts:10]: @septs I think this will not match main domain `https://pooltogether.com`

Comment (septs) [reply packages/maskbook/src/plugins/PoolTogether/constants.ts:10]:
```suggestion
export const URL_PATTERN = new RegExp(`https?:\/\/.*\.?${escapeRegExp(BASE_URL)}`)
```

File: packages/maskbook/src/plugins/PoolTogether/UI/AccountPool.tsx [hunk 26]
```diff
@@ -0,0 +1,193 @@
...
+    const classes = useStyles()
+
+    const poolURL = useManagePoolURL(accountPool.pool)
+    const chainId = useChainId()
+
+    const formatedBalance = Number.parseFloat(
+        formatBalance(accountPool.account.ticketBalance, Number.parseInt(accountPool.pool.tokens.ticket.decimals, 10)),
+    ).toLocaleString(undefined, {
+        maximumFractionDigits: 6,
+    })
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/UI/AccountPool.tsx:124]:
```suggestion
    const balance = formatBalance(
        accountPool.account.ticketBalance,
        Number.parseInt(accountPool.pool.tokens.ticket.decimals, 10),
    )
    const formattedBalance = Number.parseFloat(balance).toLocaleString(undefined, {
        maximumFractionDigits: 6
    })
```

File: packages/maskbook/src/plugins/PoolTogether/utils.ts (hunk 19 above)
Comment (septs) [reply packages/maskbook/src/plugins/PoolTogether/utils.ts]: why not use https://date-fns.org/v2.22.1/docs/intervalToDuration

Comment (septs) [reply packages/maskbook/src/plugins/PoolTogether/utils.ts]: @iRhonin

File: packages/maskbook/src/plugins/PoolTogether/utils.ts [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+export const calculateNextPrize = (pool: Pool) => {
+    const rawPrize =
+        Number.parseInt(pool.prize.weeklyTotalValueUsd, 10) /
+        (ONE_WEEK_SECONDS / Number.parseInt(pool.config.prizePeriodSeconds, 10))
+
+    let formatedPrize
+    if (!Number.isNaN(rawPrize)) {
+        formatedPrize = `\$${rawPrize.toLocaleString(undefined, { maximumFractionDigits: 0 })}`
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/utils.ts:66]:
```suggestion
        return `\$${rawPrize.toLocaleString(undefined, { maximumFractionDigits: 0 })}`
```

File: packages/maskbook/src/plugins/PoolTogether/utils.ts [hunk 28]
```diff
@@ -0,0 +1,91 @@
...
+    } else {
+        const pirzeAmount = Number.parseFloat(pool.prize.amount)
+        formatedPrize = `${pirzeAmount.toLocaleString(undefined, {
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/utils.ts:69]:
```suggestion
        return `${pirzeAmount.toLocaleString(undefined, {
```

File: packages/maskbook/src/plugins/PoolTogether/utils.ts [hunk 29]
```diff
@@ -0,0 +1,91 @@
+import { ChainId } from '@masknet/web3-shared'
+import { ONE_SECOND, ONE_WEEK_SECONDS } from './constants'
+import type { Pool } from './types'
+
+/**
+ * Converts seconds to days, hours, minutes, seconds
+ * @param totalSeconds
+ * @return {days,hours,minutes,seconds}
+ **/
+export const parseSeconds = (totalSeconds: number) => {
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/utils.ts:10]: plz use https://date-fns.org/v2.22.1/docs/intervalToDuration

File: packages/maskbook/src/plugins/PoolTogether/UI/CountdownView.tsx [hunk 30]
```diff
@@ -1,10 +1,10 @@
 import { useState } from 'react'
-import { parseSeconds } from '../utils'
 import { Box, Grid, makeStyles, Typography } from '@material-ui/core'
 import { useStylesExtends } from '../../../components/custom-ui-helper'
 import { useInterval } from 'react-use'
 import { ONE_SECOND } from '../constants'
 import { DarkColor } from '@masknet/theme/constants'
+import { intervalToDuration } from 'date-fns'
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/UI/CountdownView.tsx:7]:
```suggestion
import intervalToDuration from 'date-fns/intervalToDuration'
```

**Comment by septs** [comment]
approve, but wait for @Jack-Works and @guanbinrui  clear all problem

File: packages/maskbook/src/plugins/PoolTogether/UI/PoolView.tsx (hunk 25 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/PoolTogether/UI/PoolView.tsx:169]: I thought this is a "TODO"

**Review by Jack-Works — DISMISSED** [review]
need review from @guanbinrui and test from @AinuX

File: packages/maskbook/src/plugins/PoolTogether/UI/CountdownView.tsx [hunk 31]
```diff
@@ -0,0 +1,175 @@
...
+        return (
+            <Box py={0.2} mr={0.3} ml={0.1} className={classes.digit}>
+                <Digit digit={props.digit} />
+            </Box>
+        )
+    }
+
+    interface DoubleDigitsProps {
+        leftDigit: Number | String
+        rightDigit: Number | String
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/UI/CountdownView.tsx:107]:
```suggestion
        leftDigit: number | string
        rightDigit: number | string
```

File: packages/maskbook/src/plugins/PoolTogether/UI/CountdownView.tsx [hunk 32]
```diff
@@ -0,0 +1,175 @@
...
+    const hoursArray = hours?.toString().split('') ?? [0, 0]
+    const minutesArray = minutes?.toString().split('') ?? [0, 0]
+    const secondsArray = seconds?.toString().split('') ?? [0, 0]
+
+    // 86400 seconds = 1 day
+    // 3600 seconds = 1 hour
+    const textColor = secondsRemaining >= 86400 ? 'green-1' : secondsRemaining >= 3600 ? 'yellow-1' : 'red-1'
+
+    interface DigitProps {
+        digit: Number | String
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/UI/CountdownView.tsx:79]:
```suggestion
        digit: number | string
```

File: packages/maskbook/src/plugins/PoolTogether/UI/NetworkView.tsx [hunk 33]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+export const NetworkView = (props: NetworkViewProps) => {
+    const classes = useStylesExtends(useStyles(), props)
+    const { chainId = ChainId.Mainnet } = props
+
+    const selectedNetwork = getNetworkTypeFromChainId(chainId)
+    const chainDetail = getChainDetailed(chainId)
+
+    const color = getNetworkColor(chainDetail?.chainId ?? 1)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/UI/NetworkView.tsx:30]:
```suggestion
    const color = getNetworkColor(chainDetail?.chainId ?? ChainId.Mainnet)
```

File: packages/maskbook/src/plugins/PoolTogether/hooks/useDepositCallback.ts [hunk 34]
```diff
@@ -0,0 +1,100 @@
...
+                setDepositState({
+                    type: TransactionStateType.FAILED,
+                    error,
+                })
+                throw error
+            })
+
+        // step 2: blocking
+        return new Promise<string>((resolve, reject) => {
+            const promisEvent = poolContract.methods.depositTo(account, amount, controlledToken, referrer).send({
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/hooks/useDepositCallback.ts:71]:
```suggestion
            const promiEvent = poolContract.methods.depositTo(account, amount, controlledToken, referrer).send({
```

File: packages/maskbook/src/plugins/PoolTogether/utils.ts [hunk 35]
```diff
@@ -0,0 +1,50 @@
...
+    return startedAt + seconds - Date.now() / ONE_SECOND
+}
+
+export const getNetworkColor = (chainId: ChainId) => {
+    switch (chainId) {
+        case ChainId.Mainnet:
+            return '#617fea'
+        case ChainId.Matic:
+            return '#7b41da'
+        // add more if needed
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/PoolTogether/utils.ts:47]:
```suggestion
        case ChainId.Mainnet:
            return '#617fea'
        case ChainId.Matic:
            return '#7b41da'
        // add more if needed
        default:
            return '#f1f1f1'
```

File: packages/maskbook/src/plugins/PoolTogether/utils.ts [hunk 36]
```diff
@@ -0,0 +1,50 @@
...
+    })
+}
+
+export const calculateNextPrize = (pool: Pool) => {
+    const rawPrize =
+        Number.parseInt(pool.prize.weeklyTotalValueUsd, 10) /
+        (ONE_WEEK_SECONDS / Number.parseInt(pool.config.prizePeriodSeconds, 10))
+
+    if (!Number.isNaN(rawPrize)) {
+        return `\$${rawPrize.toLocaleString(undefined, { maximumFractionDigits: 0 })}`
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/utils.ts:26]:
```suggestion
        return '$' + rawPrize.toLocaleString(undefined, { maximumFractionDigits: 0 })
```

File: packages/maskbook/src/plugins/PoolTogether/utils.ts [hunk 37]
```diff
@@ -0,0 +1,50 @@
...
+    } else {
+        const pirzeAmount = Number.parseFloat(pool.prize.amount)
+        return `${pirzeAmount.toLocaleString(undefined, {
+            maximumFractionDigits: pirzeAmount >= 10 ? 0 : 2,
+        })} ${pool.tokens.underlyingToken.symbol}`
```
Comment (septs) [inline packages/maskbook/src/plugins/PoolTogether/utils.ts:31]:
```suggestion
        const pirzeAmount = Number.parseFloat(pool.prize.amount).toLocaleString(undefined, {
            maximumFractionDigits: pirzeAmount >= 10 ? 0 : 2,
        })
        return `${pirzeAmount} ${pool.tokens.underlyingToken.symbol}`
```

Comment (iRhonin) [reply packages/maskbook/src/plugins/PoolTogether/utils.ts:31]: Can't do that, need `prizeAmount` for formatting.

**Review by septs — APPROVED** [review]

**Comment by septs** [comment]
approve, by @guanbinrui merge, thanks

**Comment by Himself65** [comment]
```shell
packages/maskbook/src/plugins/PoolTogether/UI/PoolTogetherView.tsx:135:17 - error TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

135     } = usePool(MASK_POOL_ADDRESS, MASK_POOL_SUBGRAPH, true)
```

please fix this

**Comment by Himself65** [comment]
@guanbinrui

**Pull request merged**